File: Functions/plotResistanceIncreaseTile.m
function plotResistanceIncreaseTile(axTime, axFEC, combined_df, cell_list, cell_plot_list, label_list, colormap_list, Marker_list, plot_title)
% plotResistanceIncreaseTile - Draw a resistance-vs-time and resistance-vs-FEC
% pair of traces into caller-supplied axes, instead of creating a new figure.
%
% Summary: Mirrors the standalone plotResistanceIncrease() figure logic used
%          in PlotResistanceIncrease_Detailed.m, but targets externally
%          created axes (axTime, axFEC) so several datasets can be combined
%          as columns/tiles of one shared figure (e.g. via tiledlayout).
% Author: NEXTBMS Team
% Date: 2026-08-07
%
% Inputs:
%   axTime         - target axes for the Time [days] vs Resistance subplot
%   axFEC          - target axes for the FEC vs Resistance subplot
%   combined_df    - table with columns cell_number, Timestamp,
%                    "Resistance [Ohm]", CheckupResistanceFEC
%   cell_list      - cellstr/string array of all cell numbers, plotted as
%                    thin grey background traces for context
%   cell_plot_list - cellstr/string array of highlighted cell numbers
%   label_list     - legend label per highlighted cell
%   colormap_list  - line color per highlighted cell
%   Marker_list    - line style per highlighted cell
%   plot_title     - base title, suffixed with '(vs Time)' / '(vs FEC)'
%
% Outputs: none (draws directly into axTime/axFEC)

    % --- Time vs Resistance ---
    hold(axTime, 'on');
    % Background: every cell in the dataset, thin grey lines for context.
    for i = 1:length(cell_list)
        cell_name = cell_list{i};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        time_days = days(current_cell.("Timestamp") - current_cell.("Timestamp")(1));
        plot(axTime, time_days, current_cell.("Resistance [Ohm]"), 'LineWidth', 2, 'Color', [0.5 0.5 0.5 0.2], 'HandleVisibility', 'off');
    end

    % Highlighted cells: colored, legend-bearing traces.
    legend_handles = [];
    legend_labels = {};
    for idx = 1:length(cell_plot_list)
        cell_name = cell_plot_list{idx};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            warning('Cell %s not found in data', cell_name);
            continue;
        end
        time_days = days(current_cell.("Timestamp") - current_cell.("Timestamp")(1));
        h = plot(axTime, time_days, current_cell.("Resistance [Ohm]"), '-o', 'LineWidth', 2, 'Color', colormap_list{idx}, ...
            'MarkerFaceColor', colormap_list{idx}, 'MarkerSize', 4, 'LineStyle', Marker_list{idx});
        legend_handles = [legend_handles, h]; %#ok<AGROW>
        legend_labels{end+1} = label_list{idx}; %#ok<AGROW>
    end

    grid(axTime, 'on');
    set(axTime, 'GridLineStyle', '--', 'GridAlpha', 0.6, 'FontSize', 11);
    xlabel(axTime, 'Time [days]', 'FontSize', 11);
    ylabel(axTime, 'Checkup Resistance [\Omega]', 'FontSize', 11);
    if ~isempty(plot_title)
        title(axTime, [plot_title ' (vs Time)'], 'FontSize', 11);
    end
    if ~isempty(legend_handles)
        legend(axTime, legend_handles, legend_labels, 'FontSize', 11, 'Location', 'best', 'Interpreter', 'none');
    end
    hold(axTime, 'off');

    % --- FEC vs Resistance ---
    hold(axFEC, 'on');
    % Background: every cell in the dataset, thin grey lines for context.
    for i = 1:length(cell_list)
        cell_name = cell_list{i};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        plot(axFEC, current_cell.CheckupResistanceFEC, current_cell.("Resistance [Ohm]"), 'LineWidth', 2, 'Color', [0.5 0.5 0.5 0.2], 'HandleVisibility', 'off');
    end

    % Highlighted cells: colored, legend-bearing traces.
    legend_handles = [];
    legend_labels = {};
    maxFEC_highlight = 0;
    for idx = 1:length(cell_plot_list)
        cell_name = cell_plot_list{idx};
        mask = strcmp(combined_df.("cell_number"), cell_name);
        current_cell = combined_df(mask, :);
        if isempty(current_cell)
            continue;
        end
        h = plot(axFEC, current_cell.CheckupResistanceFEC, current_cell.("Resistance [Ohm]"), '-o', 'LineWidth', 2, 'Color', colormap_list{idx}, ...
            'MarkerFaceColor', colormap_list{idx}, 'MarkerSize', 4, 'LineStyle', Marker_list{idx});
        legend_handles = [legend_handles, h]; %#ok<AGROW>
        legend_labels{end+1} = label_list{idx}; %#ok<AGROW>
        if ~isempty(current_cell.CheckupResistanceFEC)
            maxFEC_highlight = max([maxFEC_highlight; max(current_cell.CheckupResistanceFEC)]);
        end
    end

    grid(axFEC, 'on');
    set(axFEC, 'GridLineStyle', '--', 'GridAlpha', 0.6, 'FontSize', 11);
    xlabel(axFEC, 'Full Equivalent Cycles (FEC)', 'FontSize', 11);
    ylabel(axFEC, 'Checkup Resistance [\Omega]', 'FontSize', 11);
    if ~isempty(plot_title)
        title(axFEC, [plot_title ' (vs FEC)'], 'FontSize', 11);
    end
    if maxFEC_highlight > 0
        xlim(axFEC, [0, maxFEC_highlight]);
    end
    if ~isempty(legend_handles)
        legend(axFEC, legend_handles, legend_labels, 'FontSize', 11, 'Location', 'best', 'Interpreter', 'none');
    end
    hold(axFEC, 'off');
end

File: Functions/plotResistanceIncreaseTile.py
"""
plotResistanceIncreaseTile - Draw a resistance-vs-time and resistance-vs-FEC
pair of traces (WITH legend, unlike the capacity tile) into caller-supplied
matplotlib axes.

Python counterpart of Functions/plotResistanceIncreaseTile.m, used by
python_scripts/PlotResistanceIncrease_Detailed.py (todo #019).

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-25

Compliance: R-004, R-013, R-016.
"""

_OHM_LABEL = "Checkup Resistance [$\\Omega$]"


def plot_resistance_increase_tile(ax_time, ax_fec, combined_df, cell_list,
                                   cell_plot_list, label_list, colormap_list, marker_list, plot_title):
    """
    Draw the Time [days] / FEC resistance trace pair (with a legend on each
    subplot) into ax_time/ax_fec.

    Parameters mirror plot_capacity_degradation_tile, plus label_list (one
    legend label per highlighted cell) since - unlike the capacity tile -
    this resistance tile DOES render a legend on both subplots.
    """

    # --- Time vs Resistance ---
    for cell_name in cell_list:
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        current_cell = current_cell.sort_values("Timestamp")
        time_days = (current_cell["Timestamp"] - current_cell["Timestamp"].iloc[0]).dt.total_seconds() / 86400.0
        ax_time.plot(time_days, current_cell["Resistance [Ohm]"], linewidth=2, color=(0.5, 0.5, 0.5, 0.2))

    handles, labels = [], []
    for idx, cell_name in enumerate(cell_plot_list):
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            print(f"Warning: Cell {cell_name} not found in data")
            continue
        current_cell = current_cell.sort_values("Timestamp")
        time_days = (current_cell["Timestamp"] - current_cell["Timestamp"].iloc[0]).dt.total_seconds() / 86400.0
        (handle,) = ax_time.plot(time_days, current_cell["Resistance [Ohm]"], marker="o", linewidth=2,
                                  color=colormap_list[idx], markerfacecolor=colormap_list[idx],
                                  markersize=4, linestyle=marker_list[idx])
        handles.append(handle)
        labels.append(label_list[idx])

    ax_time.grid(True, linestyle="--", alpha=0.6)
    ax_time.tick_params(labelsize=11)
    ax_time.set_xlabel("Time [days]", fontsize=11)
    ax_time.set_ylabel(_OHM_LABEL, fontsize=11)
    if plot_title:
        ax_time.set_title(f"{plot_title} (vs Time)", fontsize=11)
    if handles:
        ax_time.legend(handles, labels, fontsize=11, loc="best")

    # --- FEC vs Resistance ---
    for cell_name in cell_list:
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        ax_fec.plot(current_cell["CheckupResistanceFEC"], current_cell["Resistance [Ohm]"],
                    linewidth=2, color=(0.5, 0.5, 0.5, 0.2))

    handles, labels = [], []
    max_fec_highlight = 0.0
    for idx, cell_name in enumerate(cell_plot_list):
        current_cell = combined_df[combined_df["cell_number"] == cell_name]
        if current_cell.empty:
            continue
        (handle,) = ax_fec.plot(current_cell["CheckupResistanceFEC"], current_cell["Resistance [Ohm]"],
                                 marker="o", linewidth=2, color=colormap_list[idx],
                                 markerfacecolor=colormap_list[idx], markersize=4, linestyle=marker_list[idx])
        handles.append(handle)
        labels.append(label_list[idx])
        if current_cell["CheckupResistanceFEC"].notna().any():
            max_fec_highlight = max(max_fec_highlight, float(current_cell["CheckupResistanceFEC"].max()))

    ax_fec.grid(True, linestyle="--", alpha=0.6)
    ax_fec.tick_params(labelsize=11)
    ax_fec.set_xlabel("Full Equivalent Cycles (FEC)", fontsize=11)
    ax_fec.set_ylabel(_OHM_LABEL, fontsize=11)
    if plot_title:
        ax_fec.set_title(f"{plot_title} (vs FEC)", fontsize=11)
    if max_fec_highlight > 0:
        ax_fec.set_xlim(0, max_fec_highlight)
    if handles:
        ax_fec.legend(handles, labels, fontsize=11, loc="best")


__all__ = ["plot_resistance_increase_tile"]
